### PR TITLE
Amm

### PR DESCRIPTION
**Pull request by psolsson**
Adds AMM functionality. Unit-tests passing.

### CODE REVIEW

**Comment by franknoe** [comment]
Did you commit a vegan sausage?

File: pyemma/msm/api.py [hunk 1]
```diff
@@ -1288,6 +1292,163 @@ def bayesian_hidden_markov_model(dtrajs, nstates, lag, nsamples=100, reversible=
                                   dt_traj=dt_traj, conf=conf, store_hidden=store_hidden, show_progress=show_progress)
     return bhmsm_estimator.estimate(dtrajs)
 
+def estimate_augmented_markov_model(dtrajs, ftrajs, m, w, lag,
```
Comment (franknoe) [inline pyemma/msm/api.py]: I suggest the order ``dtrajs, ftrajs, lag, m, w`` (first simulation data and parameter, then experimental constraints)

File: pyemma/msm/api.py [hunk 2]
```diff
@@ -1288,6 +1292,163 @@ def bayesian_hidden_markov_model(dtrajs, nstates, lag, nsamples=100, reversible=
...
+    contains the estimated transition matrix and allows to compute a
+    large number of quantities related to Markov models.
+
+    Parameters
+    ----------
+    dtrajs : list containing ndarrays(dtype=int) or ndarray(n, dtype=int)
+        discrete trajectories, stored as integer ndarrays (arbitrary size)
+        or a single ndarray for only one trajectory.
+    ftrajs : list of trajectories of microscopic observables. Has to have
+        a one-to-one correspondence with the dtrajs.
```
Comment (franknoe) [inline pyemma/msm/api.py]: Has to have the same shape (number of trajectories and timesteps) as dtrajs

File: pyemma/msm/api.py [hunk 3]
```diff
@@ -1288,6 +1292,163 @@ def bayesian_hidden_markov_model(dtrajs, nstates, lag, nsamples=100, reversible=
...
+    Parameters
+    ----------
+    dtrajs : list containing ndarrays(dtype=int) or ndarray(n, dtype=int)
+        discrete trajectories, stored as integer ndarrays (arbitrary size)
+        or a single ndarray for only one trajectory.
+    ftrajs : list of trajectories of microscopic observables. Has to have
+        a one-to-one correspondence with the dtrajs. 
+    m   : Experimental averages. 
+    w   : Weight of each experimental observable. Equal to 1/(2s) where s is
+        standard error of the experimental.
```
Comment (franknoe) [inline pyemma/msm/api.py]: If we are using Gaussians, isn't this 1/(2s^2) ?
Or how about specifying s instead? Maybe this is more intuitive.

File: pyemma/msm/api.py [hunk 4]
```diff
@@ -1288,6 +1292,163 @@ def bayesian_hidden_markov_model(dtrajs, nstates, lag, nsamples=100, reversible=
...
+    m   : Experimental averages. 
+    w   : Weight of each experimental observable. Equal to 1/(2s) where s is
+        standard error of the experimental. 
+    lag : int
+        lag time at which transitions are counted and the transition matrix is
+        estimated.
+    reversible : bool, optional
+        If true compute reversible MSM, else non-reversible MSM
+    statdist : (M,) ndarray, optional
+        Stationary vector on the full state-space. Transition matrix
```
Comment (franknoe) [inline pyemma/msm/api.py]: You don't seem to have this option.

File: pyemma/msm/api.py [hunk 5]
```diff
@@ -1288,6 +1292,163 @@ def bayesian_hidden_markov_model(dtrajs, nstates, lag, nsamples=100, reversible=
...
+        *  's',   'second*'
+
+    maxiter : int, optional
+        Optional parameter with specifies the maximum number of 
+        updates for Lagrange multiplier estimation.
+         
+
+    score_method : str, optional, default='VAMP2'
+        Score to be used with MSM score function. Available scores are
+        based on the variational approach for Markov processes [13]_ [14]_:
```
Comment (franknoe) [inline pyemma/msm/api.py]: These citations seem not to be in this docstring. Do they work?

Comment (franknoe) [reply pyemma/msm/api.py]: Have you thought of whether AMM and our current cross-validation approach makes sense?

File: pyemma/msm/tests/test_amm.py [hunk 6]
```diff
@@ -0,0 +1,766 @@
+# This file is part of PyEMMA.
```
Comment (franknoe) [inline pyemma/msm/tests/test_amm.py:1]: Can't we reconcile this test with the test_msm or test_oom test? It seems we have a lot of redundant code here.

File: pyemma/msm/estimators/maximum_likelihood_msm.py [hunk 7]
```diff
@@ -1361,3 +1361,498 @@ def OOM_sigma(self):
...
+    def _update_mhat(self): 
+        """
+          Updates mhat (expectation of observable of the Augmented Markov model) 
+        """ 
+        self.mhat = _np.dot(self.pihat.reshape((self.n_mstates_active,)), self.E_active[:]) 
+        self._update_S() 
+
+    def _update_S(self):
+        """
+          Computes slope in observable space.
```
Comment (franknoe) [inline pyemma/msm/estimators/maximum_likelihood_msm.py]: In this case you can write 

""" Computes slope in observable space. """

If you have text, then use

""" Title

blabla
(empty line here)
"""

File: pyemma/msm/estimators/maximum_likelihood_msm.py [hunk 8]
```diff
@@ -1361,3 +1361,498 @@ def OOM_sigma(self):
...
+        self._is_estimated = True
+
+        # if active set is empty, we can't do anything.
+        if _np.size(self.active_set) == 0:
+            raise RuntimeError('Active set is empty. Cannot estimate AMM.')
+
+        # active count matrix and number of states
+        self._C_active = dtrajstats.count_matrix(subset=self.active_set)
+        self._nstates = self._C_active.shape[0]
+
```
Comment (franknoe) [inline pyemma/msm/estimators/maximum_likelihood_msm.py]: remove empty line (check warnings for codestyle shown on the right in PyCharm)

**Review by franknoe — CHANGES_REQUESTED** [review]

File: pyemma/msm/api.py (hunk 3 above)
Comment (psolsson) [reply pyemma/msm/api.py]: That's a typo. Thanks. Will fix.

File: pyemma/msm/api.py (hunk 1 above)
Comment (psolsson) [reply pyemma/msm/api.py]: Makes sense.

File: pyemma/msm/api.py (hunk 2 above)
Comment (psolsson) [reply pyemma/msm/api.py]: Will change.

File: pyemma/msm/api.py (hunk 5 above)
Comment (psolsson) [reply pyemma/msm/api.py]: I'm unsure whether these actually make sense. Should I remove them for now? That was my initial thought.

File: pyemma/msm/tests/test_amm.py (hunk 6 above)
Comment (psolsson) [reply pyemma/msm/tests/test_amm.py:1]: That could be possible now, but maybe it is good to keep them separate as a lot of the functionality does not translate directly between the models. Also to make room for some divergence in future developments. What do you think?

File: pyemma/msm/estimators/maximum_likelihood_msm.py (hunk 7 above)
Comment (psolsson) [reply pyemma/msm/estimators/maximum_likelihood_msm.py]: OK.

File: pyemma/msm/estimators/maximum_likelihood_msm.py (hunk 8 above)
Comment (psolsson) [reply pyemma/msm/estimators/maximum_likelihood_msm.py]: Not using PyCharm so I'm not sure what you are referring to. I'm guessing line 1683 should go?

File: pyemma/msm/tests/test_amm.py (hunk 6 above)
Comment (psolsson) [reply pyemma/msm/tests/test_amm.py:1]: I've inherited the TestMSMDoubleWell from test_msm now and overridden special cases. This removes a lot of redundancy.

**Comment by franknoe** [comment]
Cool. I think you might even be able to save most of the rest by just passing adding the AMM to the objects being tested, and only overwriting the functions starting with an "_" that actually change.

**Comment by psolsson** [comment]
That should've been done with this commit.

**Pull request merged**